### ISSUE TITLE
NoSuchMethodError with eventually in Spring Boot 2.4.x

### ISSUE THREAD

**Issue by enolive** [kotest/kotest#2265]
Kotest 4.6.0
Kotlin: 1.4.32
Spring Boot: 2.4.5

I noticed that `eventually` does not seem to be compatible any longer with the current Spring Boot version. When calling a test that uses `eventually`, I get a `NoSuchMethodError`

> java.lang.NoSuchMethodError: 'java.lang.Object io.kotest.assertions.timing.EventuallyKt.eventually-D5N0EJY(double, kotlin.jvm.functions.Function1, kotlin.coroutines.Continuation)'

This happens both with the classic method and with the new one from kotest.frramwork.concurrency

build.gradle.kts

```kotlin
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

repositories {
  mavenCentral()
}

plugins {
  id("org.springframework.boot") version "2.4.5"
  id("io.spring.dependency-management") version "1.0.11.RELEASE"
  kotlin("jvm") version "1.4.32"
  kotlin("plugin.spring") version "1.4.32"
}

group = "com.example"
version = "0.0.1-SNAPSHOT"
java.sourceCompatibility = JavaVersion.VERSION_11

val kotest_version = "4.6.0"

dependencies {
  implementation("org.springframework.boot:spring-boot-starter-webflux")
  implementation("com.fasterxml.jackson.module:jackson-module-kotlin")
  implementation("io.projectreactor.kotlin:reactor-kotlin-extensions")
  implementation("org.jetbrains.kotlin:kotlin-reflect")
  implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
  implementation("org.jetbrains.kotlinx:kotlinx-coroutines-reactor")
  testImplementation("org.springframework.boot:spring-boot-starter-test")
  testImplementation("io.projectreactor:reactor-test")
  testImplementation("io.kotest:kotest-runner-junit5:$kotest_version")
  testImplementation("io.kotest:kotest-assertions-core:$kotest_version")
  testImplementation("io.kotest:kotest-property:$kotest_version")
  testImplementation("io.kotest.extensions:kotest-extensions-spring:1.0.0")
  testImplementation("com.ninja-squad:springmockk:3.0.1")
}

tasks.withType<KotlinCompile> {
  kotlinOptions {
    freeCompilerArgs = listOf("-Xjsr305=strict")
    jvmTarget = "11"
  }
}

tasks.withType<Test> {
  useJUnitPlatform()
}
```

ApplicationTest.kt:

```kotlin
package com.example.demo

import io.kotest.assertions.timing.eventually
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.nulls.shouldNotBeNull
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.core.env.Environment
import kotlin.time.ExperimentalTime
import kotlin.time.seconds

@ExperimentalTime
@SpringBootTest
class ApplicationTest(private val environment: Environment) : DescribeSpec({
  describe("application") {
    it("has a working environment") {
      environment.shouldNotBeNull()
    }

    it("works eventually") {
      eventually(2.seconds) {
        true.shouldBeFalse()
      }
    }
  }
})
```

As a side note, the next Spring Boot snapshot version with Kotlin 1.5.0 works perfectly. 

Any chance for a fix or should I wait for the next Spring Boot release before upgrading?

**Comment by sksamuel** [comment]
I would recommend you migrate to the new eventually module in 4.6 which doesn't use durations in the signatures so shouldn't give you any errors whether on 1.4 or 1.5
See https://kotest.io/docs/framework/concurrency/eventually.html

**Comment by enolive** [comment]
unfortunately, I get the same error message when using eventually from k`kotest-framework-concurrency`

Did I miss something in the docs?

**Comment by sksamuel** [comment]
@jschneidereit any ideas?

This package shouldn't use durations at all. You need to use longs as the duration time, not the Duration type.

**Comment by enolive** [comment]
after watching the [release notes of Kotlin 1.5](https://www.youtube.com/watch?v=LpqvtgibbsQ&t=620s), I can share some additional information on this problem with you

TLDR: 

The `Duration` inline value class previously had a Double internal value, which was changed to Long to allow more precision. As value classes are inlined, this is unfortunately a breaking change for functions that use this class. What might be expected for experimental APIs, though :wink: 

That's why both the old and new APIs in kotest produce a NoSuchMethodError (you might notice the double type in the error message).

I might be wrong, but I see no way of fixing that but to upgrade to Kotlin 1.5 (other than switching to the primitive version of `eventually`). Maybe a documentation hint would be nice that the functions that use `Duration` require Kotlin 1.5

**Comment by sksamuel** [comment]
Yep, the internal representation changed, from Double to Long as you say, which is why we introduced the new eventually module, to avoid these errors. The new implementation doesn't use the Duration class (other than an overload which you're not using).

See the code here:
https://github.com/kotest/kotest/blob/master/kotest-framework/kotest-framework-concurrency/src/commonMain/kotlin/io/kotest/framework/concurrency/eventually.kt

So if you bring this module into your build, and use the new packages, you should see this error go away. Note the new package name of `io.kotest.framework.concurrency` to distinguish it from the existing eventually code.

**Comment by sksamuel** [comment]
Did this fix it @enolive ?

**Comment by enolive** [comment]
sorry for the delayed reply. I tried the new API and it works, at least when using the builder variant!

what irritates me, I cannot make the config variant (which might be great for reusing a configuration as described in the docs) work:

```kotlin
it("fails eventually with builder") {
      eventually({
        duration = 5_000
      }) {
        true.shouldBeFalse()
      }
    }
    it("fails eventually with config") {
      eventually(EventuallyConfig(duration = 5_000)) {
        true.shouldBeFalse()
      }
    }
```

instead of retrying, the second version will immediately return the result. Did I miss something in the docs?
